Add post "Creating a Python Virtual Environment for QGIS Plugin Development with VS Code on Windows" (#10)
geotribu/english-blog: content/posts/2024-11-25_pyqgis_environnement_dev_windows.md

---
authors:
  - nicogodet
categories:
  - Tutorial
comments: true
date: 2024-11-25
description: "For Intellisense happiness"
icon: material/microsoft-visual-studio-code

> license: beerware
> pin: false

Reviewer: ```suggestion
license: beerware
links:
    - "Cooking with Gispo: QGIS Plugin Development in VS Code": https://www.gispo.fi/en/blog/qgis-plugin-development-in-vs-code/
    - IDE settings for writing and debugging plugins (official PyQGIS cookbook): https://docs.qgis.org/3.34/en/docs/pyqgis_developer_cookbook/plugins/ide_debugging.html
    - French version: https://geotribu.fr/articles/2024/2024-11-25_pyqgis_environnement_dev_windows/
pin: false
```

tags:
    - Plugin QGIS
    - PyQGIS
    - Python
    - VS Code
    - Windows

> title: "Creating a Python Virtual Environment for QGIS Plugin Development with VS Code on Windows"

Reviewer: ```suggestion
title: "Creating a Python virtual environment for PyQGIS development with VS Code on Windows."
```

subtitle: Keep our env safe, for our PyQGIS children
---

> # Creating a Python Virtual Environment for QGIS Plugin Development with VS Code on Windows

Reviewer: ```suggestion
# Creating a Python virtual environment for developing QGIS plugin with VS Code on Windows
```

## Introduction

![PyQGIS logo](https://cdn.geotribu.fr/img/logos-icones/programmation/pyqgis.png){: .img-thumbnail-left }

> Anyone who has tried it knows that configuring a Python, PyQGIS, and PyQt environment on Windows for developing QGIS plugins is a real challenge. Often, it feels like a losing battle...

Reviewer: ```suggestion
Anyone who has tried it, knows that configuring a Python, PyQGIS, and PyQt environment on Windows for developing QGIS plugins is a real challenge. Often, it feels like a losing battle...
```

> Well, not anymore! After scouring the depths of the internet and exploring tips provided by [Julien](https://geotribu.fr/team/julien-moura/), here is one method to have (almost) all the auto-completions for PyQGIS, PyQt, and more in VS Code.

Reviewer: ```suggestion
Well, not anymore! After scouring the depths of the internet and exploring tips provided by [Julien](https://geotribu.fr/team/julien-moura/), here is one method to have (almost) all the auto-completions for PyQGIS, PyQt and more in VS Code.
```

<!-- more -->

[Leave a comment :fontawesome-solid-comments:](#__comments "Go to comments"){: .md-button }
{: align=middle }

----

## Creating the Virtual Environment

I will assume that you have installed QGIS in the `C:\OSGeo4W` directory (the procedure is the same whether QGIS is installed via the OSGeo4W network installer or the MSI package; the paths in this article just need to be adapted to your installation).

1. Open an OSGeo4W Shell and navigate to the location where you want to create the virtual environment.  
   For example, a freshly created plugin template using the [QGIS Plugin Templater](https://gitlab.com/Oslandia/qgis/template-qgis-plugin).

1. Run the following commands:

    ```cmd title="Creating a virtual environment in the OSGeo4W Shell"
    C:\OSGeo4W\bin\python-qgis.bat -m venv --system-site-packages .venv
    C:\OSGeo4W\bin\python-qgis.bat -c "import pathlib;import qgis;print(str((pathlib.Path(qgis.__file__)/'../..').resolve()))" > .venv\qgis.pth
    ```

    The `--system-site-packages` option allows the virtual environment to inherit libraries specific to the Python environment in QGIS.

> 1. To ensure VSCode recognizes `processing` imports, add the following line to the `.venv\qgis.pth` file:  

Reviewer: ```suggestion
1. To ensure VS Code recognizes `processing` imports, add the following line to the `.venv\qgis.pth` file:  
```

    `C:\OSGeo4W\apps\qgis\python\plugins`

    Your file should look like this:

>     ```text title="Contents of .venv\qgis.pth file"

Reviewer: ```suggestion
    ```text title="Content of .venv\qgis.pth file"
```
